### PR TITLE
add support for studio trace id in selectors and document context entry

### LINKED ISSUE #3803: docs: add `apollo_operation_id`
In #3586, @o0Ignition0o introduced the `apollo_operation_id` — a user-accessible attribute that can be fetched from the context and used to build a durable link to the Operation page in Studio.   This is useful for emitting into logs or, e.g., as an attribute on trace spans for debugging purposes.  We may make this available in more places in the future!

Using two other pieces of information from the graph reference (the `your-graph@current` option that you run your Router with), this new `apollo_operation_id` value can be used to generate the full URL using the following convention:

```
https://studio.apollographql.com/graph/{{your_graph_id}}/variant/{{your_graph_variant}}/operations?query={{apollo_operation_id}}
```

Where:

* `{{your_graph_id}}` (This is the part to the left of the `@` sign in your "graph reference" (i.e., the `your-graph` part of `your-graph@current`)
* `{{your_graph_variant}}` (This is the part to the right of the `@` sign in your "graph reference" (i.e., the `current` part of `your-graph@current`)
* `{{apollo_operation_id}}` this is the value read from the context's `apollo_operation_id` property.

No curly braces should be present in the actual URL.  😄

### LINKED ISSUE #5172: Add support for studio trace id selector in telemetry
We already have support for dd trace id an otel trace id. We should add studio one

### PR DESCRIPTION
**Pull request by bnjjj**
Add support for a new trace id selector kind, the apollo trace id which represents the trace id on [Apollo Studio](https://studio.apollographql.com/). 

Example:

```yaml
telemetry:
  instrumentation:
    spans:
      router:
        "studio.trace.id":
            trace_id: apollo
```

Add documentation for available rhai constants.

Fixes #3803 
Fixes #5172

---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [x] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [x] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

File: docs/source/customizations/rhai-api.mdx [hunk 1]
```diff
@@ -337,6 +337,19 @@ fn router_service(service) {
 
 </Note>
 
+## Available constants
+
+You can access to provided constants, it's mainly context keys to use when you want to fetch data from the context.
```
Comment (garypen) [inline docs/source/customizations/rhai-api.mdx:342]:
```suggestion
You can access provided constants, it's mainly context keys to use when you want to fetch data from the context.
```

**Review by garypen — CHANGES_REQUESTED** [review]
I'm requesting change for the selectors.mdx change because I don't think that will work properly without escaping the pipe.
Apart from that it looks great and thanks for documenting the Rhai constants.

**Comment by bnjjj** [comment]
@garypen good catch ! Thanks

**Review by garypen — APPROVED** [review]

File: docs/source/customizations/rhai-api.mdx [hunk 2]
```diff
@@ -337,6 +337,19 @@ fn router_service(service) {
 
 </Note>
 
+## Available constants
+
+You can access provided constants, it's mainly context keys to use when you want to fetch data from the context.
```
Comment (shorgi) [inline docs/source/customizations/rhai-api.mdx:342]:
```suggestion
The router provides constants for your Rhai scripts that mostly help you fetch data from the context.
```

File: docs/source/configuration/telemetry/apollo-telemetry.mdx [hunk 3]
```diff
@@ -331,3 +331,10 @@ telemetry:
           account: # Override the default behavior for the "account" subgraph
             send: false
 ```
+
+
+<Note>
+
+If you're writing a plugin you can get the Studio Trace ID by reading the value of `apollo_operation_id` from the context.
```
Comment (shorgi) [inline docs/source/configuration/telemetry/apollo-telemetry.mdx:338]:
```suggestion
If you're writing a plugin, you can get the Studio Trace ID by reading the value of `apollo_operation_id` from the context.
```

File: .changesets/feat_bnjjj_add_studio_trace_id_selector.md [hunk 4]
```diff
@@ -0,0 +1,18 @@
+### Add support for studio trace id in selectors and document context entry ([Issue #3803](https://github.com/apollographql/router/issues/3803)), ([Issue #5172](https://github.com/apollographql/router/issues/5172))
+
+Add support for a new trace id selector kind, the apollo trace id which represents the trace id on [Apollo Studio](https://studio.apollographql.com/). 
+
+Example:
```
Comment (shorgi) [inline .changesets/feat_bnjjj_add_studio_trace_id_selector.md:5]: Do separate issues require separate changelog entries? Can edit in the eventual release's changelog.

```suggestion
### Add support for studio trace id in selectors and document context entry ([Issue #3803](https://github.com/apollographql/router/issues/3803)), ([Issue #5172](https://github.com/apollographql/router/issues/5172))

Add support for a new trace ID selector kind, the `apollo` trace ID, which represents the trace ID on [Apollo GraphOS Studio](https://studio.apollographql.com/). 

An example configuration using `trace_id: apollo`:
```

File: .changesets/feat_bnjjj_add_studio_trace_id_selector.md [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+```yaml
+telemetry:
+  instrumentation:
+    spans:
+      router:
+        "studio.trace.id":
+            trace_id:: apollo
+```
+
+Add documentation for available rhai constants.
```
Comment (shorgi) [inline .changesets/feat_bnjjj_add_studio_trace_id_selector.md:16]:
```suggestion
Also, router documentation has been updated with constants that are available in Rhai.
```

File: apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap [hunk 6]
```diff
@@ -6241,6 +6241,13 @@ expression: "&schema"
             "datadog"
           ],
           "type": "string"
+        },
+        {
+          "description": "Apollo Studio trace id",
```
Comment (shorgi) [inline apollo-router/src/configuration/snapshots/apollo_router__configuration__tests__schema_generation.snap:6246]:
```suggestion
          "description": "Apollo Studio trace ID",
```

**Review by shorgi — COMMENTED** [review]
Docs review

**Review by BrynCooke — COMMENTED** [review]
Can we introduce an integration test for this? I think it should jsut work as the apollo key and graph ref is present.

**Comment by bnjjj** [comment]
@BrynCooke integration test added

File: docs/source/customizations/rhai-api.mdx [hunk 7]
```diff
@@ -337,6 +337,19 @@ fn router_service(service) {
...
+
+```
+Router.APOLLO_SDL // Context key to access the SDL
+Router.APOLLO_START // Constant to calculate durations
+Router.APOLLO_AUTHENTICATION_JWT_CLAIMS // Context key to access authentication jwt claims
+Router.APOLLO_SUBSCRIPTION_WS_CUSTOM_CONNECTION_PARAMS // Context key to modify or access the custom connection params when using subscriptions in WebSocket to subgraphs (cf subscription docs)
+Router.APOLLO_ENTITY_CACHE_KEY // Context key to access the entity cache key
+Router.APOLLO_OPERATION_ID // Context key to get the value of apollo operation id (studio trace id) from the context
+```
+
```
Comment (abernix) [inline docs/source/customizations/rhai-api.mdx:352]: @shorgi Not urgent, but this probably needs some other visual treatment.

**Comment by abernix** [comment]
I think this whole feature should have been called "operation ID"?  We pluck the value from `APOLLO_OPERATION_ID`; it's **not** a trace ID, right?

**Pull request merged**